bundles/org.openhab.binding.mqtt/src/main/java/org/openhab/binding/mqtt/discovery/AbstractMQTTDiscovery.java

```diff
@@ -36,25 +40,61 @@
  */
 @NonNullByDefault
 public abstract class AbstractMQTTDiscovery extends AbstractDiscoveryService implements MQTTTopicDiscoveryParticipant {
+
+    private final Logger logger = LoggerFactory.getLogger(AbstractMQTTDiscovery.class);
+
     protected final String subscribeTopic;
 
+    private int timeout;
+
+    private @Nullable ScheduledFuture<?> scheduledStop;
+
     public AbstractMQTTDiscovery(@Nullable Set<ThingTypeUID> supportedThingTypes, int timeout,
             boolean backgroundDiscoveryEnabledByDefault, String baseTopic) {
-        super(supportedThingTypes, timeout, backgroundDiscoveryEnabledByDefault);
+        super(supportedThingTypes, 0, backgroundDiscoveryEnabledByDefault);
         this.subscribeTopic = baseTopic;
+        this.timeout = timeout;
     }
 
     /**
      * Return the topic discovery service.
      */
     protected abstract MQTTTopicDiscoveryService getDiscoveryService();
 
+    private synchronized void stopTimeout() {
+        if (scheduledStop != null) {
+            scheduledStop.cancel(false);
+            scheduledStop = null;
+        }
+    }
+
+    protected synchronized void resetTimeout() {
+        stopTimeout();
+
+        // schedule an automatic call of stopScan when timeout is reached
+        if (timeout > 0) {
+            Runnable runnable = new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        stopScan();
+                    } catch (Exception e) {
```

Comment:
Is it really necessary to catch Exception here? What exception do you expect`?

Reply:
see above: the code is basically taken from `org.eclipse.smarthome.config.discovery.AbstractDiscoveryService`. They thought it would be good tp do it, so I took their example.

Reply:
Eclipse smarthome contains a lot of over engineered, pre Java-8 code that is only slowly refactored and converted into modern Java 11 code. I suggest not taking it as example code.